starknet_py/net/gateway_client.py

```diff
@@ -0,0 +1,328 @@
+
+from starkware.starknet.definitions.fields import ContractAddressSalt
+from starkware.starkware_utils.error_handling import StarkErrorCode
+
+
+from starknet_py.transaction_exceptions import (
+    TransactionFailedError,
+    TransactionRejectedError,
+    TransactionNotReceivedError,
+)
+from starknet_py.net.base_client import (
+    BaseClient,
+    BlockHashIdentifier,
+    BlockNumberIdentifier,
+)
+from starknet_py.net.client_models import (
+    Transaction,
+    SentTransaction,
+    ContractCode,
+    TransactionReceipt,
+    BlockState,
+    StarknetBlock,
+    InvokeFunction,
+    StarknetTransaction,
+    ContractDefinition,
+    Deploy,
+    TransactionStatus,
+)
+from starknet_py.net.gateway_schemas.gateway_schemas import (
+    TransactionSchema,
+    ContractCodeSchema,
+    StarknetBlockSchema,
+    TransactionReceiptSchema,
+    SentTransactionSchema,
+)
+from starknet_py.net.models import StarknetChainId, chain_from_network
+from starknet_py.net.networks import Network, net_address_from_net
+
+
+class GatewayClient(BaseClient):
+    def __init__(self, net: Network, chain: StarknetChainId = None):
+        host = net_address_from_net(net)
+        feeder_gateway_url = f"{host}/feeder_gateway"
+        gateway_url = f"{host}/gateway"
+
+        self.chain = chain_from_network(net, chain)
+        self._feeder_gateway_client = StarknetClient(url=feeder_gateway_url)
+        self._gateway_client = StarknetClient(url=gateway_url)
+
+    async def get_transaction(
+        self,
+        tx_identifier: Union[int, BlockHashIdentifier, BlockNumberIdentifier],
+    ) -> Transaction:
+        if isinstance(tx_identifier, (BlockHashIdentifier, BlockNumberIdentifier)):
+            raise ValueError(
+                "BlockHashIdentifier and BlockNumberIdentifier are not supported in gateway client."
+            )
+        res = await self._feeder_gateway_client.call(
+            method_name="get_transaction",
+            params={
+                "transactionHash": str(hex(tx_identifier))
+                if isinstance(tx_identifier, int)
+                else tx_identifier
+            },
+        )
+        # TODO handle not received/unknown transactions
+        return TransactionSchema().load(res["transaction"], unknown=EXCLUDE)
+
+    async def get_block(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> StarknetBlock:
+        block_identifier = self._get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+
+        res = await self._feeder_gateway_client.call(
+            method_name="get_block", params=block_identifier
+        )
+        return StarknetBlockSchema().load(res, unknown=EXCLUDE)
+
+    async def get_state_update(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> BlockState:
+        # TODO implement
```

Comment:
Should we support get_state_update if it wasn't supported in previous client?

Reply:
Why not?

Reply:
Sure, guess there's not reason for not supporting it now